### PR TITLE
Sb 3715- Content meta filters Outline and testcase for review

### PR DESCRIPTION
**Pull request by revathitarento**
Content meta filters Outline and test case, previous review comments are also implemented.

### CODE REVIEW

File: src/middlewares/filter.middleware.js [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+  // If the request body has filter by metaFilter data, continue with the same filter, do not alter the values
+
+  // else call the getMetaSearchString() function to generate the search string for the meta filters
+  filterService.getMetaSearchString(function () {
+    // Generate JSON and return
+    return setFilterJSON()
+  })
+}
+
+function setFilterJSON () {
```
Comment (kochhar) [inline src/middlewares/filter.middleware.js]: This additional callback function may not be required.

File: src/service/filterService.js [hunk 2]
```diff
@@ -6,9 +6,9 @@
  */
 var configUtil = require('sb-config-util')
 
-function getChannelSearchString (callback) {
-  var searchString = configUtil.getConfig('CHANNEL_FILTER_QUERY_STRING')
+function getMetaSearchString (callback) {
```
Comment (kochhar) [inline src/service/filterService.js]: this should probably be renamed to not indicate a string.

File: src/test/middlewares/requestMiddlewareSpec.js [hunk 3]
```diff
@@ -51,3 +51,10 @@ describe('filter of channels', function () {
     })
   })
 })
+
+describe('Initialization of meta filters', function () {
```
Comment (kochhar) [inline src/test/middlewares/requestMiddlewareSpec.js]: these tests should probably move to `filterMiddlewareSpec.js`

File: src/test/channelFilterRouteSpec.js [hunk 4]
```diff
@@ -65,3 +65,7 @@ describe('Check for all route to be calling the AddChannelFilter', function () {
     })
   })
 })
+describe('Check for all route to be calling the AddMetaFilter', function () {
```
Comment (kochhar) [inline src/test/channelFilterRouteSpec.js]: these tests need to be generalised to go beyond just channel filters.

File: src/middlewares/filter.middleware.js [hunk 5]
```diff
@@ -0,0 +1,16 @@
+var _ = require('underscore')
+var filterService = require('../service/filterService')
+
+function addMetaFilters (req, res, next) {
+  // If the request body has filter by metaFilter data, continue with the same filter, do not alter the values
+  // else call the getMetaSearchData() function to generate the JSON Object for the meta filters
+  filterService.getMetaSearchData(function () {
+    // Generate JSON and return
+    return setFilterJSON()
```
Comment (kochhar) [inline src/middlewares/filter.middleware.js]: change `set` => `get`

File: src/middlewares/request.middleware.js [hunk 6]
```diff
@@ -295,24 +295,24 @@ function checkChannelID (req, res, next) {
...
-      } else if (channels && (!_.isEmpty(channels))) {
-        req.body.request.filters.channel = channels
-        LOG.info(utilsService.getLoggerData({}, 'INFO',
-          filename, 'addChannelFilters', 'added channel filter', req.body.request.filters.channel))
-      }
-      next()
-    })
-  }
-}
+// function addChannelFilters (req, res, next) {
```
Comment (kochhar) [inline src/middlewares/request.middleware.js]: Please document this function with *some* explanation of what this code will do.

**Comment by revathitarento** [comment]
@kochhar Few more test scenarios are updated

File: src/app.js [hunk 7]
```diff
@@ -38,6 +38,12 @@ const learnerServiceLocalBaseUrl = process.env.sunbird_learner_service_local_bas
 
 const whiteListedChannelList = process.env.sunbird_content_service_whitelisted_channels
 const blackListedChannelList = process.env.sunbird_content_service_blacklisted_channels
+const whitelistedframeworkList = process.env.sunbird_content_service_whitelisted_framework
+const blacklistedframeworkList = process.env.sunbird_content_service_blacklisted_framework
+const whitelistedLanguageList = process.env.sunbird_content_service_whitelisted_language
```
Comment (kochhar) [inline src/app.js]: Based on design we should be considering `contentType` and `mimeType` instead of language and creator.

File: src/app.js [hunk 8]
```diff
@@ -124,7 +130,8 @@ if (defaultChannel) {
           'start service Eg: sunbird_environment = dev, sunbird_instance = sunbird')
           process.exit(1)
         }
-        updateConfig(getFilterConfig())
+        // updateConfig(getFilterConfig())
+        configUtil.setConfig('META_FILTER_QUERY_STRING', getMetaFilterConfig())
```
Comment (kochhar) [inline src/app.js]: var name needs to be updated.

File: src/app.js [hunk 9]
```diff
@@ -165,26 +172,22 @@ const telemetryConfig = {
...
-  }
-  LOG.info(utilsService.getLoggerData({}, 'INFO',
-    filename, 'getFilterConfig', 'config string', configString))
-  return configString
+function getFilterJSONFromEnv () {
+  // Generate JSON and return
+  return null
+}
+
+function getFilterJSONfromConfigService () {
```
Comment (kochhar) [inline src/app.js]: this should raise an exception since it is not implemented.

File: src/service/filterService.js [hunk 10]
```diff
@@ -6,9 +6,9 @@
  */
 var configUtil = require('sb-config-util')
 
-function getChannelSearchString (callback) {
-  var searchString = configUtil.getConfig('CHANNEL_FILTER_QUERY_STRING')
-  callback(null, searchString)
+function getMetaSearchData (callback) {
```
Comment (kochhar) [inline src/service/filterService.js]: suggest renaming to `getMetadataFilterQuery`

File: src/service/filterService.js [hunk 11]
```diff
@@ -6,9 +6,9 @@
...
+  var searchData = configUtil.getConfig('META_FILTER_REQUEST_JSON')
```
Comment (kochhar) [inline src/service/filterService.js]: The var does not match the var which was set.

Comment (kochhar) [inline src/service/filterService.js]: suggest renaming to `filterQuery`

File: src/test/service/contentServiceSpec.js [hunk 12]
```diff
@@ -113,6 +113,12 @@ describe('Composite search services', function () {
     })
     done()
   })
+
+  it('check for metafilter in config with whitelist', function (done) {
```
Comment (kochhar) [inline src/test/service/contentServiceSpec.js]: not sure what this test is testing. could you add some details?

File: src/test/service/contentServiceSpec.js [hunk 13]
```diff
@@ -113,6 +113,12 @@ describe('Composite search services', function () {
...
+  })
+
+  it('check for metafilter in config with blacklist', function (done) {
```
Comment (kochhar) [inline src/test/service/contentServiceSpec.js]: same as above.

File: src/middlewares/request.middleware.js [hunk 14]
```diff
@@ -295,24 +295,26 @@ function checkChannelID (req, res, next) {
...
+//           'failed to get channels'))
+//       } else if (channels && (!_.isEmpty(channels))) {
+//         req.body.request.filters.channel = channels
+//         LOG.info(utilsService.getLoggerData({}, 'INFO',
+//           filename, 'addChannelFilters', 'added channel filter', req.body.request.filters.channel))
+//       }
+//       next()
+//     })
+//   }
+// }
```
Comment (rajeevsathish) [inline src/middlewares/request.middleware.js]: Please remove the commented code don't keep it commented

File: src/app.js (hunk 7 above)
Comment (revathitarento) [reply src/app.js]: @kochhar The changes are done on different branch, and i tried to push to this current branch. So this push has outdated code. I will merge the changes and push the code.

File: src/app.js (hunk 8 above)
Comment (revathitarento) [reply src/app.js]: will be updated

File: src/middlewares/filter.middleware.js [hunk 15]
```diff
@@ -3,15 +3,14 @@ var filterService = require('../service/filterService')
 
 function addMetaFilters (req, res, next) {
   // If the request body has filter by metaFilter data, continue with the same filter, do not alter the values
-
-  // else call the getMetaSearchString() function to generate the search string for the meta filters
-  filterService.getMetaSearchString(function () {
+  // else call the getMetaSearchData() function to generate the JSON Object for the meta filters
+  filterService.getMetaSearchData(function () {
     // Generate JSON and return
     return setFilterJSON()
```
Comment (kochhar) [inline src/middlewares/filter.middleware.js]: Change `set` => `get`

File: src/middlewares/filter.middleware.js [hunk 16]
```diff
@@ -3,15 +3,14 @@ var filterService = require('../service/filterService')
 
 function addMetaFilters (req, res, next) {
   // If the request body has filter by metaFilter data, continue with the same filter, do not alter the values
-
-  // else call the getMetaSearchString() function to generate the search string for the meta filters
-  filterService.getMetaSearchString(function () {
+  // else call the getMetaSearchData() function to generate the JSON Object for the meta filters
+  filterService.getMetaSearchData(function () {
```
Comment (kochhar) [inline src/middlewares/filter.middleware.js]: does this need to be a callback based method?

File: src/service/filterService.js (hunk 11 above)
Comment (kochhar) [reply src/service/filterService.js]: Please fix this.

Comment (kochhar) [reply src/service/filterService.js]: Please fix this.

File: src/test/service/contentServiceSpec.js (hunk 12 above)
Comment (revathitarento) [reply src/test/service/contentServiceSpec.js]: Details added

File: src/test/service/contentServiceSpec.js (hunk 13 above)
Comment (revathitarento) [reply src/test/service/contentServiceSpec.js]: Details added

File: src/service/filterService.js (hunk 11 above)
Comment (revathitarento) [reply src/service/filterService.js]: fixed

File: src/service/filterService.js (hunk 10 above)
Comment (revathitarento) [reply src/service/filterService.js]: Done

File: src/middlewares/request.middleware.js (hunk 14 above)
Comment (revathitarento) [reply src/middlewares/request.middleware.js]: Removed the commented code

File: src/middlewares/request.middleware.js (hunk 6 above)
Comment (revathitarento) [reply src/middlewares/request.middleware.js]: removed the function as it is not called now

File: src/middlewares/filter.middleware.js (hunk 16 above)
Comment (revathitarento) [reply src/middlewares/filter.middleware.js]: Not required, functionality has been implemented and tested locally, it is working fine.

File: src/app.js (hunk 8 above)
Comment (revathitarento) [reply src/app.js]: updated

File: src/app.js (hunk 7 above)
Comment (revathitarento) [reply src/app.js]: updated

**Pull request closed without merging**